src/lazy/binary/lazy_reader.rs

```diff
@@ -0,0 +1,222 @@
+use crate::binary::constants::v1_0::IVM;
+use crate::element::reader::ElementReader;
+use crate::element::Element;
+use crate::lazy::binary::system::lazy_system_reader::LazySystemReader;
+use crate::lazy::binary::system::lazy_value::LazyValue;
+use crate::result::{decoding_error, decoding_error_raw};
+use crate::IonResult;
+
+/// A binary reader that only reads each value that it visits upon request (that is: lazily).
+///
+/// Each time [`LazyReader::next`] is called, the reader will advance to the next top-level value
+/// in the input stream. Once positioned on a top-level value, users may visit nested values by
+/// calling [`LazyValue::read`] and working with the resulting [`crate::lazy::value_ref::ValueRef`],
+/// which may contain either a scalar value or a lazy container that may itself be traversed.
+///
+/// The values that the reader yields ([`LazyValue`],
+/// [`crate::lazy::binary::system::lazy_sequence::LazySequence`], and
+/// [`crate::lazy::binary::system::lazy_struct::LazyStruct`]) are
+/// immutable references to the data stream, and remain valid until [`LazyReader::next`] is called
```

Comment:
🗺️ Unfortunately, importing a type to make the doc comment more readable like this:
```rust
use crate::lazy::binary::system::lazy_sequence::LazySequence;

/// This is a doc comment that refers to [`LazySequence`].
```
is not supported. The compiler will complain about an unused import.